File: mobile-apps/alloy-android/app/src/main/kotlin/com/alloyremote/alloy/onboarding/OnboardingFlow.kt
package com.alloyremote.alloy.onboarding

import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.platform.LocalContext

/**
 * Top-level onboarding navigation. Owns the `OnboardingViewModel` and
 * routes between the six wizard screens defined in
 * `specs/onboarding.md`.
 */
@Composable
fun OnboardingFlow() {
    val context = LocalContext.current
    val model = remember { OnboardingViewModel(context) }
    val step by model.step.collectAsState()

    when (val s = step) {
        OnboardingViewModel.Step.Welcome -> WelcomeScreen(onStart = { model.start() })

        OnboardingViewModel.Step.Discovering -> DiscoveryScreen(
            model = model,
            onConnectHA = { model.connectHomeAssistant() },
            onContinue = { model.continueToPicker() }
        )

        OnboardingViewModel.Step.ConnectingHomeAssistant -> HomeAssistantConnectScreen(model = model)

        OnboardingViewModel.Step.Picking -> TargetPickerScreen(
            model = model,
            onSelect = { avr -> model.selectForTest(avr) }
        )

        is OnboardingViewModel.Step.Testing -> TestConnectionScreen(
            avr = s.avr,
            model = model,
            onConfirm = {
                val descriptor = model.descriptorFor(s.avr)
                model.finish(descriptor)
            }
        )

        is OnboardingViewModel.Step.Done -> {
            // RootScreen will switch away once bound target lands in
            // settings. This branch is a momentary bridge.
        }
    }
}

File: mobile-apps/alloy-ios/Alloy/Onboarding/OnboardingFlow.swift
import SwiftUI

/// Top-level onboarding navigation. Owns the `OnboardingViewModel`
/// and routes between the six wizard screens defined in
/// `specs/onboarding.md`.
struct OnboardingFlow: View {
    @State private var model = OnboardingViewModel()
    @Environment(SettingsStore.self) private var settings

    var body: some View {
        NavigationStack {
            content
                .navigationTitle("")
                .navigationBarTitleDisplayMode(.inline)
        }
    }

    @ViewBuilder
    private var content: some View {
        switch model.step {
        case .welcome:
            WelcomeView { model.start() }

        case .discovering:
            DiscoveryView(
                combiner: model.combiner,
                onConnectHA: { model.connectHomeAssistant() },
                onContinue: { model.continueToPicker() }
            )

        case .connectingHomeAssistant:
            HomeAssistantConnectView(model: model)

        case .picking:
            TargetPickerView(
                combiner: model.combiner,
                onSelect: { avr in model.selectForTest(avr) }
            )

        case .testing(let avr):
            TestConnectionView(
                avr: avr,
                model: model,
                onConfirm: {
                    let descriptor = model.descriptor(for: avr)
                    settings.boundTargetDescriptor = descriptor
                    model.finish(descriptor)
                }
            )

        case .done:
            // Root view switches away as soon as `boundTargetDescriptor`
            // lands in settings. This case is just a momentary bridge.
            ProgressView()
        }
    }
}
